core/src/main/java/org/apache/iceberg/hadoop/HadoopFileIO.java

```diff
@@ -63,7 +63,11 @@ public class HadoopFileIO implements HadoopConfigurable, DelegateFileIO {
    * <p>{@link Configuration Hadoop configuration} must be set through {@link
    * HadoopFileIO#setConf(Configuration)}
    */
-  public HadoopFileIO() {}
+  public HadoopFileIO() {
+    // Create a default hadoopConf as it is required for the object to be valid.
+    // E.g. newInputFile would throw NPE with hadoopConf.get() otherwise.
+    this.hadoopConf = new SerializableConfiguration(new Configuration())::get;
```

Comment:
FileIOParser doesn't serialize and deserialize the Hadoop configuration. the deserialized io object is not valid because `hadoopConf` is null. NPE would be throw if `newInputFile` is called on deserialized object.

not sure if we want to change `FileIOParser` to serialize Hadoop `Configuration` object or not. Regardless, I feel this change seems reasonable, as this class requires `hadoopConf` to not be null.

Reply:
I thought about this problem a bit more. The one downside I see with this change is that `HadoopFileIO` will typically be instantiated via `CatalogUtil.loadFileIO`, which will then set the `Conf`. Also the javadoc of the constructor mentions that the conf must be set by calling `setConf()`.

Instead of serializing the conf in the Parser, maybe the conf should be initialized in `FileIOParser.fromJson(String)` instead of passing a null conf in https://github.com/apache/iceberg/blob/7fe4037e778583489ccdf6e06a78645384d4b9f3/core/src/main/java/org/apache/iceberg/io/FileIOParser.java#L68? I think that would solve the issue while also not instantiating a new Conf every time a new instance of `HadoopFileIO` is created?

Reply:
if the `FileIO` is `Configurable`, create a default `Configuration` object  and call `setConf()`. is that what you mean? that will add Hadoop class dependency to the `FileIOParser` though.

I have also thought about implementing the serialization of Hadoop `Configuration` in the `FileIOParser`. but it will add some complexity to the `FileIOParser` and we likely won't achieve 100% fidelity with the original object. E.g., we may only serialize the key-value string pairs from the configuration as a JSON object. 

I also thought about use this method from `SerializationUtil`? But it is Java serialization (not JSON serialization).
```
  /**
   * Serialize an object to bytes. If the object implements {@link HadoopConfigurable}, its Hadoop
   * configuration will be serialized into a {@link SerializableConfiguration}.
   *
   * @param obj object to serialize
   * @return serialized bytes
   */
  public static byte[] serializeToBytes(Object obj) {
    return serializeToBytes(obj, conf -> new SerializableConfiguration(conf)::get);
  }
```

Reply:
ah you're right that this will add a Hadoop dependency and we don't want that. 
Rather than setting a conf in the constructor, maybe we should have some handling in `conf()`, such as 
```
   public Configuration conf() {
-    return hadoopConf.get();
+    return Optional.ofNullable(hadoopConf).map(Supplier::get).orElseGet(Configuration::new);
   }
```

We would need to adjust all the places that call `hadoopConf.get()` to call `conf()` after this change. That way we wouldn't need to unnecessarily instantiate a conf in the constructor when we don't have to, wdyt?

Reply:
`conf()` method is fine. 

`FileIOParser` doesn't faithfully serialize and deserialize the `HadoopFileIO` (Hadoop configuration is not carried over). The deserialized `HadoopFileIO` may miss important configs, which can be a problem.

from `CatalogUtil`, there are 3 components defining a `FileIO`. `FileIOParser` is missing the `conf`. 
```
FileIO loadFileIO(String impl, Map<String, String> properties, Object hadoopConf)
```

I am wondering if we should change `FileIOParser` to serialize and deserialize Hadoop `Configuration` when the `FileIO` is `HadoopConfigurable`. We can probably only serialize the key-value string pairs from the `Configuration` as a JSON object (kind of a read only copy).


BTW, `ResolvingFileIO` and `HadoopConfigurable` also have Hadoop class dependency. There was a discussion of potentially moving `HadoopCatalog` to a separate `iceberg-hadoop` module. I guess we can't move `ResolvingFileIO` and `HadoopConfigurable` then.

Reply:
FYI there's a similar NPE issue in `ResolvingFileIO` that I looked at (https://github.com/apache/iceberg/pull/10872) but in the `ResolvingFileIO` we should probably default to a `null` config while here it seems to make more sense to create a new config if the configured one is null.

> I am wondering if we should change FileIOParser to serialize and deserialize Hadoop Configuration when the FileIO is HadoopConfigurable. We can probably only serialize the key-value string pairs from the Configuration as a JSON object (kind of a read only copy).

Yes I had similar thoughts on this topic. Let me think about this for a bit and we can discuss how we'd like to approach/improve this situation

Reply:
> I am wondering if we should change FileIOParser to serialize and deserialize Hadoop Configuration when the FileIO is HadoopConfigurable. We can probably only serialize the key-value string pairs from the Configuration as a JSON object (kind of a read only copy).

I am also a bit undecided on this topic, but was also wondering whether we should ser/de the Hadoop config in the `FileIOParser`. I did a quick check and I think something like the below might work, where we ser/de the actual bytes.
`SerializationUtil` currently takes the entire `FileIO` object rather than just the `Configuration` in the below example, but if we would want to go down that route we can most likely improve this.

```
--- a/core/src/main/java/org/apache/iceberg/io/FileIOParser.java
+++ b/core/src/main/java/org/apache/iceberg/io/FileIOParser.java
@@ -21,10 +21,16 @@ package org.apache.iceberg.io;
 import com.fasterxml.jackson.core.JsonGenerator;
 import com.fasterxml.jackson.databind.JsonNode;
 import java.io.IOException;
+import java.nio.ByteBuffer;
 import java.util.Map;
 import org.apache.iceberg.CatalogUtil;
+import org.apache.iceberg.hadoop.HadoopConfigurable;
+import org.apache.iceberg.hadoop.HadoopFileIO;
 import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.io.BaseEncoding;
+import org.apache.iceberg.util.ByteBuffers;
 import org.apache.iceberg.util.JsonUtil;
+import org.apache.iceberg.util.SerializationUtil;

 public class FileIOParser {
   private FileIOParser() {}
@@ -61,6 +67,12 @@ public class FileIOParser {
     generator.writeStringField(FILE_IO_IMPL, impl);
     JsonUtil.writeStringMap(PROPERTIES, properties, generator);

+    if (io instanceof HadoopFileIO) {
+      ByteBuffer wrap = ByteBuffer.wrap(SerializationUtil.serializeToBytes(io));
+      generator.writeStringField(
+          "hadoopConf", BaseEncoding.base16().encode(ByteBuffers.toByteArray(wrap)));
+    }
+
     generator.writeEndObject();
   }

@@ -76,6 +88,14 @@ public class FileIOParser {
     Preconditions.checkArgument(json.isObject(), "Cannot parse FileIO from non-object: %s", json);
     String impl = JsonUtil.getString(FILE_IO_IMPL, json);
     Map<String, String> properties = JsonUtil.getStringMap(PROPERTIES, json);
-    return CatalogUtil.loadFileIO(impl, properties, conf);
+    FileIO fileIO = CatalogUtil.loadFileIO(impl, properties, conf);
+
+    if (json.hasNonNull("hadoopConf") && fileIO instanceof HadoopFileIO) {
+      ByteBuffer buffer = JsonUtil.getByteBufferOrNull("hadoopConf", json);
+      HadoopConfigurable o = SerializationUtil.deserializeFromBytes(buffer.array());
+      HadoopFileIO hadoopFileIO = (HadoopFileIO) fileIO;
+      hadoopFileIO.setConf(o.getConf());
+    }
+    return fileIO;
```

Reply:
I have two concerns for the above approach

1. it is not a JSON serialization of the Hadoop config. it is Java serialization. Plus, we are serializing the whole `FileIO` and store the bytes as `hadoopConf`, which conceptually is not correct.
2. it introduces class dependency to `HadoopFileIO` in the `FileIOParser`.

Reply:
I am leaning toward implementing Hadoop configuration JSON serialization as string key-value pairs. The only down side is the dependency to Hadoop class in `FileIOParser`.

Another alternative is to merge Hadoop configuration key-value pairs into the `Map<String, String>` properties. `setConf` would also load the Hadoop config key-value pairs with lower-priority into the properties map. map properties are also passed along with higher override property to Hadoop configuration. this could avoid dependency to Hadoop class in `FileIOParser`.

 @rdblue @Fokko any feedback too?

Reply:
I'm very concerned about the approach to serialize/deserialize the Hadoop Configuration. There could potentially be hundreds, sometimes thousands of config keys in a Configuration objects (I have seen above 2 thousands in some cases). When we start writing out string key-value parts, we even lose the gzip compression we currently have for the Configuration.write method. This will increase the size of the object seriously.

I think we should consider some ways to reduce the size of the objects:
1. Use empty config - if the config is really used this will cause issues
2. Serialize only the non-default values - version change could cause issues
3. Use the one provided by the catalog - task/job specific configs can cause issues

I'm leaning towards the 2nd option

Reply:
@Fokko: Sorry, I missed your comment. Previously we considered a `PlainParquetConfiguration` like solution, and we decided against it at that time, because it would have needed a big sweeping change for minimal gains at that time. Maybe with the "hadoop independent" PRs committed we are in a better shape now?

Reply:
> There could potentially be hundreds, sometimes thousands of config keys in a Configuration objects (I have seen above 2 thousands in some cases).

agree that size is non-trivial. but note that `FileIO` is not serialized per file scan task. it was [serialized per manifest file scan task](https://github.com/apache/iceberg/pull/10735/files#diff-8bc8bee8f0eb396c2e6bf7175fca3207fad4daead009d8cc25aa2360f8544c55). So the impact is not too bad.

> 1. Use empty config - if the config is really used this will cause issues

this is what the current PR is doing. but it is not correct. Hadoop configuration is part of the `HadoopFileIO` object state and `FileIOParser` is carrying it over during serialization.

> Serialize only the non-default values - version change could cause issues

Is this doable? As you were also saying, it can be problematic, as the object deserialization also depends on the environment. that is also technically incorrect.

> 3. Use the one provided by the catalog - task/job specific configs can cause issues

`FileIOParser` has no access to a catalog. Plus, it also could be a `RESTCatalog` with `HadoopFileIO`.

> This way we can serialize Hadoops Configuration into a Map<String, String> and deserialize it into Parquets' PlainParquetConfiguration equivalent, which is probably much more lightweight.

@Fokko I want to point out that this change/fix is for the iceberg-core module, not just for Flink. Let me make sure that I understand you correctly. 

`FileIOParser` can serialize the configuration as string key-value pairs, that is inline with what I was saying earlier about serialization only string key-value pairs. 

But we want to avoid using Hadoop `Configuration` class to set. Then we probably need to extend Iceberg's `HadoopConfigurable` interface with a new setter, like `setConfigPropertes(Map<String, String> configProperties`.  Then why don't we merge the Hadoop configuration with the existing `Map<String, String>` properties in the `HadoopFileIO`, which was the [alternative](https://github.com/apache/iceberg/pull/10926#discussion_r1733692980) approach I mentioned in this comment.

Reply:
If we need to do the FileIO serialization, then we will add a serious amount of data for the tasks:
```
  @Test
  public void testA() throws IOException {
    Configuration confWithDefaults = new Configuration();
    Configuration confWithoutDefaults = new Configuration(false);
    SerializableConfiguration sconfWithDefaults = new SerializableConfiguration(confWithDefaults);
    SerializableConfiguration sconfWithoutDefaults = new SerializableConfiguration(confWithoutDefaults);
    byte[] sconfWithDefaultBytes = SerializationUtil.serializeToBytes(sconfWithDefaults);
    byte[] sconfWithoutDefaultBytes = SerializationUtil.serializeToBytes(sconfWithoutDefaults);
    String jsonWithDefaults = String.join(", " , Lists.newArrayList(confWithDefaults.iterator()).stream().map(e -> e.getKey() + ":" + e.getValue()).collect(Collectors.toList()));
    ByteArrayOutputStream out = new ByteArrayOutputStream();
    DataOutputViewStreamWrapper view = new DataOutputViewStreamWrapper(out);
    view.writeUTF(jsonWithDefaults);
    byte[] jsonWithDefaultsBytes = out.toByteArray();
    String jsonWithoutDefaults = String.join(", " , Lists.newArrayList(confWithDefaults.iterator()).stream().map(e -> e.getKey() + ":" + e.getValue()).collect(Collectors.toList()));
    out = new ByteArrayOutputStream();
    view = new DataOutputViewStreamWrapper(out);
    view.writeUTF(jsonWithoutDefaults);
    byte[] jsonWithoutDefaultsBytes = out.toByteArray();
    System.out.println("Binary serialization length with defaults: " + sconfWithDefaultBytes.length);
    System.out.println("Binary serialization length without defaults: " + sconfWithoutDefaultBytes.length);
    System.out.println("JSON serialization length with defaults: " + jsonWithDefaultsBytes.length);
    System.out.println("JSON serialization length without defaults: " + jsonWithoutDefaultsBytes.length);
  }
```
Results in:
```
Binary serialization length with defaults: 68638
Binary serialization length without defaults: 76
JSON serialization length with defaults: 36282
JSON serialization length without defaults: 36282
```

I have seen that we use `new Configuration(false)` in the code, so we allow for the user to provide a trimmed configuration, and in this case the serialized config is quite small for the binary serialization. We might have to do something similar for the JSON serialization.

We use the serialized FileIO for:
- ManifestListReadTask
- ManifestReadTask

If we step back one more time. We might consider why do we need the FileIO in the `ManifestListReadTask` at all?
The `ManifestListReadTask.rows()` and `ManifestListReadTask.file()` is using the `io` to get the new input file like `io.newInputFile(manifestListLocation)`. I think this FileIO serialization will become even more complicated when the manifest file encryption arrives here. We will need to apply the encryption for the FileIO like this:
```
          EncryptingFileIO.combine(table().io(), table().encryption())
```

We could serialize/deserialize the DataFile to return the `ManifestListReadTask.file()` like we do it for the `BaseFileScanTask`, but I don't see a better option for the `ManifestListReadTask.rows()` yet 😢

Reply:
> I have seen that we use new Configuration(false) in the code, so we allow for the user to provide a trimmed configuration, and in this case the serialized config is quite small for the binary serialization. We might have to do something similar for the JSON serialization.

Are you saying trim the key-value pairs with the entries from default configuration as `new Configuration(false)`. We can potentially do that. but it still has the implication that it depends on the runtime env. if the other side (deserialization) has a different env (default config), this can be different.

> The ManifestListReadTask.rows() and ManifestListReadTask.file() is using the io to get the new input file like io.newInputFile(manifestListLocation). 

`FileIO` is used to read from manifest file. `ManifestFiles.read` is a widely used API.
```
    private CloseableIterable<? extends ContentFile<?>> files(Schema fileProjection) {
      switch (manifest.content()) {
        case DATA:
          return ManifestFiles.read(manifest, io, specsById).project(fileProjection);
        case DELETES:
          return ManifestFiles.readDeleteManifest(manifest, io, specsById).project(fileProjection);
...
    }
```

>  FileIO serialization will become even more complicated when the manifest file encryption arrives here. We will need to apply the encryption for the FileIO 

Looking at the usage of `EncryptingFileIO`, it is not part of the table ops state. It should never need to be serialized. only original `FileIO` and `EncryptionManager` need to be serialized.

```
  protected EncryptedOutputFile newManifestOutputFile() {
    String manifestFileLocation =
        ops.metadataFileLocation(
            FileFormat.AVRO.addExtension(commitUUID + "-m" + manifestCount.getAndIncrement()));
    return EncryptingFileIO.combine(ops.io(), ops.encryption())
        .newEncryptingOutputFile(manifestFileLocation);
  }
```

Reply:
> but it still has the implication that it depends on the runtime env. if the other side (deserialization) has a different env (default config), this can be different.

Yes, but I think (not tested) we suffer from the same issue, if the configuration provided through the Catalog is created as `Configuration(false)`.

> FileIO is used to read from manifest file. ManifestFiles.read is a widely used API.

Yes, but this case the reader provides the FileIO independently from the Task.

> only original FileIO and EncryptionManager need to be serialized.

Yes. Encryption should be handled

Reply:
@pvary if I understand correctly, your point is that `FileIO` probably shouldn't be part of the task state for `BaseEntriesTable.ManifestReadTask`, `AllManifestsTable.ManifestListReadTask`? Then we don't need to serialize `FileIO` for those manifest related tasks. That would be a fair question. It seems like a large/challenging refactoring. Looking for other folks' take on this. 

Regardless, the issue remains that `FileIOParser` doesn't serialize `HadoopFileIO` faithfully. I don't know if REST catalog has any need to use it to JSON serialize `FileIO` in the future.

Reply:
@stevenzwu: 
> if I understand correctly, your point is that FileIO probably shouldn't be part of the task state for BaseEntriesTable.ManifestReadTask, AllManifestsTable.ManifestListReadTask? [..] It seems like a large/challenging refactoring. Looking for other folks' take on this.

Yes, this seems strange, and problematic to me. I was not able to find an easy solution yet. I was hoping, others with better knowledge might have some ideas.

> Regardless, the issue remains that FileIOParser doesn't serialize HadoopFileIO faithfully. I don't know if REST catalog has any need to use it to JSON serialize FileIO in the future.

My point here is that, since we use `Configuration(false)` in some cases, and the way how the current serialization works, we already `doesn't serialize HadoopFileIO faithfully`. So If we don't find a solution for getting rid of the FileIO, we might as well write our own "unfaithful" serialization which mimics the way how the current serialization works.

Reply:
@pvary coming back to the size concern of serializing Hadoop Configuration, that is already the case with Java serialization for the manifest tasks. Implementing JSON serialization for Hadoop configuration is not making things worse.

whether we can avoid the need of `FileIO` in the manifest tasks can be a separate larger discussion.